knimin/lib/format.py

```diff
@@ -0,0 +1,273 @@
+                detail = {'lane': lane, 'run_name': run_name + str(idx)}
+                sample_details.append(sample_detail_fmt % detail)
+        else:
+            sample_details = [sample_detail_fmt % {'run_name': run_name}]
+    else:
+        sample_details = [sample_detail_fmt % samp
+                          for samp in sample_information]
+
+    base_sheet = _format_general(run_name, labadmin_id, assay, fwd_cycles,
+                                 rev_cycles, pi_name, pi_email,
+                                 contact_0_name, contact_0_email,
+                                 c1name, c1email,
+                                 c2name, c2email)
+
+    full_sheet = "%s%s\n%s\n" % (base_sheet, sample_header,
+                                 '\n'.join(sample_details))
+
+    return full_sheet
+
+
+def _format_general(run_name, labadmin_id, assay, fwd_cycles, rev_cycles,
+                    pi_name, pi_email,
+                    contact_0_name, contact_0_email,
+                    contact_1_name=None, contact_1_email=None,
+                    contact_2_name=None, contact_2_email=None):
+    """Format the initial parts of a sample sheet
+
+    Parameters
+    ----------
+    run_name : str
+        The name of the run.
+    labadmin_id : int
+        The ID of the sequencing run as informed by labadmin
+    assay : str
+        The assay instrument (e.g., Kapa Hyper Plus)
+    fwd_cycles : int
+        The number of forward cycles
+    rev_cycles : int
+        The number of reverse cycles
+    pi_name : str
+        The name of the principle investigator
+    pi_email : str
+        The email address of the principle investigator
+    contact_0_name : str
+        The name of an additional contact person
+    contact_0_email : str
+        The email of an additional contact person
+    contact_1_name : str, optional
+        The name of an additional contact.
+    contact_1_email : str, optional
+        The email of an additional contact.
+    contact_2_name : str, optional
+        The name of an additional contact.
+    contact_2_email : str, optional
+        The email of an additional contact.
+
+    Raises
+    ------
+    ValueError
+        If any required entries are None or the empty string
+
+    Returns
+    -------
+    str
+        The populated non-sample parts of the sample sheet.
+    """
+    fmt = {
+        'run_name': run_name,
+        'assay': assay,
+        'date': datetime.datetime.now().strftime("%m/%d/%Y"),
+        'fwd_cycles': fwd_cycles,
+        'rev_cycles': rev_cycles,
+        'labadmin_id': labadmin_id,
+        'pi_name': pi_name,
+        'pi_email': pi_email,
+        'contact_0_name': contact_0_name,
+        'contact_0_email': contact_0_email
+    }
+
+    optional = {
+        'contact_1_name': contact_1_name,
+        'contact_1_email': contact_1_email,
+        'contact_2_name': contact_2_name,
+        'contact_2_email': contact_2_email
+    }
+
+    for k, v in fmt.items():
```

Comment:
Not important, I just wonder if this long function can be simplified by things like `locals` and `exec` with a for loop.